lib/ops/events/codes.go

```diff
@@ -0,0 +1,430 @@
+		Name: AppUninstalledEvent,
+		Code: ApplicationUninstallCode,
+	}
+)
+
+const (
+	// OpereationInstallStartCode is the install operation start event code.
+	OperationInstallStartCode = "G0001I"
+	// OperationInstallCompleteCode is the install operation complete event code.
+	OperationInstallCompleteCode = "G0002I"
+	// OperationInstallFailureCode is the install operation failure event code.
+	OperationInstallFailureCode = "G0002E"
+	// OperationExpandStartCode is the expand operation start event code.
+	OperationExpandStartCode = "G0003I"
+	// OperationExpandCompleteCode is the expand operation complete event code.
+	OperationExpandCompleteCode = "G0004I"
+	// OperationExpandFailureCode is the expand operation failure event code.
+	OperationExpandFailureCode = "G0004E"
+	// OperationShrinkStartCide is the shrink operation start event code.
+	OperationShrinkStartCode = "G0005I"
+	// OperationShrinkCompleteCode is the shrink operation complete event code.
+	OperationShrinkCompleteCode = "G0006I"
+	// OperationShrinkFailureCode is the shrink operation failure event code.
+	OperationShrinkFailureCode = "G0006E"
+	// OperationUpdateStartCode is the update operation start event code.
+	OperationUpdateStartCode = "G0007I"
+	// OperationUpdateCompeteCode is the update operation complete event code.
+	OperationUpdateCompleteCode = "G0008I"
+	// OperationUpdateFailureCode is the update operation failure event code.
+	OperationUpdateFailureCode = "G0008E"
+	// OperationUninstallStartCode is the uninstall operation start event code.
+	OperationUninstallStartCode = "G0009I"
+	// OperationUninstallCompleteCode is the uninstall operation complete event code.
+	OperationUninstallCompleteCode = "G0010I"
+	// OperationUninstallFailureCode is the uninstall operation failure event code.
+	OperationUninstallFailureCode = "G0010E"
+	// OperationGCStartCode is the garbage collect operation start event code.
+	OperationGCStartCode = "G0011I"
+	// OperationGCCompleteCode is the garbage collect operation complete event code.
+	OperationGCCompleteCode = "G0012I"
+	// OperationGCFailureCode is the garbage collect operation failure event code.
+	OperationGCFailureCode = "G0012E"
+	// OperationEnvStartCode is the runtime environment update operation start event code.
+	OperationEnvStartCode = "G0013I"
+	// OperationEnvCompleteCode is the runtime environment update operation complete event code.
+	OperationEnvCompleteCode = "G0014I"
+	// OperationEnvFailureCode is the runtime environment update operation failure event code.
+	OperationEnvFailureCode = "G0014E"
+	// OperationConfigStartCode is the cluster configuration update operation start event code.
+	OperationConfigStartCode = "G0015I"
+	// OperationConfigCompleteCode is the cluster configuration update operation complete event code.
+	OperationConfigCompleteCode = "G0016I"
+	// OperationConfigFailureCode is the cluster configuration update operation failure event code.
+	OperationConfigFailureCode = "G0016E"
```

Comment:
I assume there's logic behind the numbering scheme - can you elaborate? Is there a reason not to keep the sub-groups in the same number cluster or why are certain values skipped?

Reply:
Yeah, I tried to maintain some logic just for consistency for these events, e.g.

Operation events are in `00xx` group,
Resource-related events (or things that are created/deleted in general) are in `1xxx` (for "create") and `2xxx` (for "delete") groups,
Cluster-related events (remote support, status ,etc.) are in `3xxx`,
Application-related events are in `4xxx`.

I'll add a comment about it - there's no strict rules about choosing codes (e.g. for future events), I just tried to setup some basic rules for these ones.